### PR TITLE
Darlin serializations

### PR DESCRIPTION
**Pull request by DanieleDiBenedetto**

### CODE REVIEW

**Review by phoinic — APPROVED** [review]

File: src/ahp/prover.rs [hunk 1]
```diff
@@ -62,14 +62,35 @@ pub enum ProverMsg<F: Field> {
 }
 
 impl<F: Field> algebra::ToBytes for ProverMsg<F> {
-    fn write<W: std::io::Write>(&self, w: W) -> std::io::Result<()> {
+    fn write<W: Write>(&self, mut writer: W) -> std::io::Result<()> {
         match self {
-            ProverMsg::EmptyMessage => Ok(()),
-            ProverMsg::FieldElements(field_elems) => field_elems.write(w),
+            ProverMsg::EmptyMessage => 0u32.write(&mut writer),
```
Comment (UlrichHaboeck75) [inline src/ahp/prover.rs:67]: We write a u32 zero for an empty message? If these four bytes (really four?) encode the length of the message than that's ok.

File: src/data_structures.rs [hunk 2]
```diff
@@ -25,23 +30,17 @@ pub struct VerifierKey<F: PrimeField, PC: PolynomialCommitment<F>>
...
-impl<F: PrimeField, PC: PolynomialCommitment<F>> Clone
-    for VerifierKey<F, PC>
-{
-    fn clone(&self) -> Self {
-        Self {
-            index_comms: self.index_comms.clone(),
-            index_info: self.index_info.clone(),
-        }
+impl<F: PrimeField, PC: PolynomialCommitment<F>> algebra::SemanticallyValid for VerifierKey<F, PC> {
+    fn is_valid(&self) -> bool {
```
Comment (UlrichHaboeck75) [inline src/data_structures.rs:42]: We could widen the notion of validity of a verifier key beyond just commitment checks, by some general checks on the index_info.

File: src/data_structures.rs [hunk 3]
```diff
@@ -71,17 +73,10 @@ pub struct ProverKey<
...
+impl<F: PrimeField, PC: PolynomialCommitment<F>> algebra::SemanticallyValid for ProverKey<F, PC>
 {
-    fn clone(&self) -> Self {
-        Self {
-            index_vk: self.index_vk.clone(),
-            index_comm_rands: self.index_comm_rands.clone(),
-            index: self.index.clone(),
-        }
+    fn is_valid(&self) -> bool {
+        self.index_vk.is_valid()
```
Comment (UlrichHaboeck75) [inline src/data_structures.rs:79]: The same here: we could add the checks that the arithmetizations of the R1CS matrices are correct, or at least do some lightweight checks as degree bounds, etc.

File: src/data_structures.rs [hunk 4]
```diff
@@ -102,6 +98,103 @@ pub struct Proof<F: PrimeField, PC: PolynomialCommitment<F>> {
     pub pc_proof: PC::BatchProof,
 }
 
+impl<F: PrimeField, PC: PolynomialCommitment<F>> algebra::SemanticallyValid for Proof<F, PC> {
+    fn is_valid(&self) -> bool {
+        self.commitments.is_valid() &&
+            self.evaluations.is_valid() &&
+            self.pc_proof.is_valid()
```
Comment (UlrichHaboeck75) [inline src/data_structures.rs:105]: I expect that the checks here are just elementwise. We could do also some additional ones such as on the number of commitments and evaluations.

**Review by UlrichHaboeck75 — COMMENTED** [review]
As mentioned in my comments, we could elaborate on the notion of semantic validity of key material, etc., by some further cheap checks on metrics (the number of commitments, evaluations, the index info, etc.)

File: src/ahp/prover.rs [hunk 5]
```diff
@@ -61,6 +61,15 @@ pub enum ProverMsg<F: Field> {
...
 
+impl<F: Field> SemanticallyValid for ProverMsg<F> {
+    fn is_valid(&self) -> bool {
+        match self {
+            ProverMsg::EmptyMessage => true,
+            ProverMsg::FieldElements(fes) => fes.is_valid()
+        }
+    }
+}
+
```
Comment (UlrichHaboeck75) [inline src/ahp/prover.rs:72]: Shouldn't we assume throughout that ``ProverMsg`` is empty?

**Review by UlrichHaboeck75 — APPROVED** [review]
Very well written update!

File: src/data_structures.rs [hunk 6]
```diff
@@ -26,23 +31,30 @@ pub struct VerifierKey<F: PrimeField, PC: PolynomialCommitment<F>>
...
-impl<F: PrimeField, PC: PolynomialCommitment<F>> algebra::ToBytes
-    for VerifierKey<F, PC>
-{
-    fn write<W: std::io::Write>(&self, mut w: W) -> std::io::Result<()> {
-        self.index_info.write(&mut w)?;
-        self.index_comms.write(&mut w)
+impl<F: PrimeField, PC: PolynomialCommitment<F>> algebra::ToBytes for VerifierKey<F, PC> {
+    #[inline]
+    fn write<W: Write>(&self, writer: W) -> std::io::Result<()> {
+        self.index_comms.serialize_uncompressed(writer)
```
Comment (albertog78) [inline src/data_structures.rs:37]: index_info ?

**Review by albertog78 — CHANGES_REQUESTED** [review]

**Review by albertog78 — APPROVED** [review]

**Pull request merged**